### PR TITLE
ENH: extend pairing options

### PR DESCRIPTION
**Pull request by aburrell**
Extend the possible model-instrument pairing options for models with a wide range of applicable times and locations.

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

This has been tested by pairing up DINEOFs ExB drifts for a day with JRO ExB drifts measured over the course of a day.  The figure below shows the instrument and model results for a day and all altitudes.

jro_drifts_slt_comp_22apr14

**Test Configuration**:
* Operating system: OS X Sierra
* Version number: 3.7

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by rstoneback** [comment]
DINEOFs looks to be producing data with a healthy offset

**Comment by aburrell** [comment]
Yup.  The good thing is that the local time dependence is very similar (except for the PRE, which is documented to have a larger altitude dependence).

**Comment by rstoneback** [comment]
That is true. All negative daytime drifts though...

When I checked against JRO in previous years the C/NOFS comparison was much better. That data was closest to JRO, low altitude, and not run through DINEOFs. Not exactly the same but I was hoping those qualities would come through better.

I'm in the process of grabbing the last version of C/NOFS IVM data made by Robin so I can reprocess it to include the new magnetic unit vectors. I'll need to go through offset process and such as well. When I get through all of that I'll be able to run some altitude dependence plots as well as do a new run with DINEOFs. Hopefully that improves things. I'll also switch to my so called 'medium' resolution based on Sarah's comments on the call.

I could potentially include a curl-free ion drift condition on the DINEOF processing if none of that helps.

**Comment by aburrell** [comment]
See if you can't get the zonal drift offsets working well with the new magnetic unit vectors.  That may be useful for getting appropriate altitude variations worked in.

**Comment by aburrell** [comment]
Also, any comments on the PR?

**Comment by rstoneback** [comment]
> Also, any comments on the PR?

Sure, I'm surprised isn't showing up more, or really at all. It is in the averaged seasonal data.

I can offer general guidance. We are pushing very low data coverage percentages at the moment. Can only do so much.

Including the geophysical boundary condition does offer more information to the system, it may help.

There was also mention of other data sets that I could ingest.

**Comment by rstoneback** [comment]
> See if you can't get the zonal drift offsets working well with the new magnetic unit vectors. That may be useful for getting appropriate altitude variations worked in.

I can only get meridional drift offsets. It should help getting appropriate altitude variations worked in. Shouldn't hurt to use the correct vectors.....

I'm starting with Robin's latest RPA processing which will have more impact on the zonal. RPA does contribute to meridional at times.

**Comment by aburrell** [comment]
> Sure, I'm surprised isn't showing up more, or really at all. It is in the averaged seasonal data.
PR = Pull Request, not PRE.

**Comment by rstoneback** [comment]
My bad. On the list for today :)

File: pysatModels/utils/extract.py [hunk 1]
```diff
@@ -627,10 +647,24 @@ def extract_modelled_observations(inst, model, inst_name, mod_name,
...
+    # and model data.  If only one time value is present, assume anything is
+    # close enough
+    dtime = mod_datetime[1:] - mod_datetime[:-1]
+    if len(dtime) == 0:
+        tm_sec = np.inf
+    else:
+        tm_sec = dmod_time.min()
+        tm_sec /= np.timedelta64(1, 's')
+
+    dtime = inst.index[1:] - inst.index[:-1]
```
Comment (rstoneback) [inline pysatModels/utils/extract.py]: I've been tripped up with pandas formats a few times where it auto aligns things by time before doing the op and I end up with a bunch of zeros. I may be paranoid, but when taking deltas I tend to toss a .values in there to get back to a numpy array.

File: pysatModels/utils/extract.py [hunk 2]
```diff
@@ -627,10 +647,24 @@ def extract_modelled_observations(inst, model, inst_name, mod_name,
...
                                   mod_datetime_name]))
 
-    tm_sec = (mod_datetime[1:] - mod_datetime[:-1]).min()
-    tm_sec /= np.timedelta64(1, 's')
-    ti_sec = (inst.index[1:] - inst.index[:-1]).min().total_seconds()
-    min_del = tm_sec if tm_sec < ti_sec else ti_sec
+    # Determine the appropriate time difference in seconds from the instrument
+    # and model data.  If only one time value is present, assume anything is
+    # close enough
+    dtime = mod_datetime[1:] - mod_datetime[:-1]
```
Comment (rstoneback) [inline pysatModels/utils/extract.py]: Similar comment here as given below.

Comment (aburrell) [reply pysatModels/utils/extract.py]: Not needed here, since these were assigned using values

**Pull request merged**